### ISSUE TITLE
[bug] moon run --affected not working as moon ci (affected not gathered) 

### ISSUE THREAD

**Issue by spljs** [moonrepo/moon#952]
**Describe the bug**

The command `moon ci` is doing great by getting the touched files . However we can't achieve the same result for an individual task like `moon run --affected :test`

```
moon run  --affected :test                                       

OUTPUT:

Target(s) :test not affected by touched files (using status all)
```

**Steps to reproduce**

1. Create two libs
2. Create a global task like :test
3. Git add , commit everything
4. Run moon ci (working)
5. Run moon run --affected :test (Not working ! Not doing anything)

**Expected behavior**

To run the task on affected files like on `moon ci`

**Comment by milesj** [comment]
@spljs If the `moon run --affected` is running in CI, you'll also need to pass `--remote` I believe.

**Comment by spljs** [comment]
Unfortunately , i've forgot to mention but already tried with the `--remote` flag, and it does the same :(

**Comment by milesj** [comment]
@spljs What's the log output? `--remote` uses `git diff`, otherwise it uses `git status`.

**Comment by spljs** [comment]
```
+ moon run --affected --remote -c 1 :test
Target(s) :test not affected by touched files (using status all)
▪▪▪▪ There's a new version of moon available, 1.9.1 (currently on 1.9.0)!
▪▪▪▪ VCS hooks, smarter task inheritance, and more!
▪▪▪▪ Run moon upgrade or install from https://moonrepo.dev/docs/install
latest: Pulling from python-poetry-moon-arm
Digest: sha256:xxxxxxxx
Status: Image is up to date for xxxxxxx/python-poetry-moon-arm:latest
+ moon run --affected --remote -c 1 :test
Target(s) :test not affected by touched files (using status all)
```

Here is the CI log but it outputs the same on local.

1. I'm modifying a project file 
2. Git add & commit
3. Testing on local with `moon ci` (gathering the right touched files)
4. Testing on local with `moon run --affected --remote -c 1 :test` Output: Target(s) :test not affected by touched files (using status all)
5. Testing on CI with the same command above (same result)

**Comment by spljs** [comment]
Here is the output of `moon ci`

```                                
▪▪▪▪ Gathering touched files
  .drone.yml
  libs/mylib/mypackage/main.py
▪▪▪▪ Gathering runnable targets
xxxxxxx:test

▪▪▪▪ Generating dependency graph
Target count: 1
Action count: 10
▪▪▪▪ Running all targets

```

**Comment by milesj** [comment]
Try passing `--log trace` as well

**Comment by spljs** [comment]
I can not paste the complete log for confidential issues but here are the git operations with 

`moon run -c 1 --affected --remote :test`

```
[DEBUG 23:45:26] moon_process::command_inspector Running command git merge-base master HEAD env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:26] moon_process::command_inspector Running command git merge-base origin/master HEAD env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:26] moon_process::command_inspector Running command git merge-base upstream/master HEAD env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:26] moon_process::command_inspector Running command git merge-base HEAD master origin/master env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:26] moon_process::command_inspector Running command git --no-pager diff --name-status --no-color --relative -z 50a02c3967485af89a9e8c1f08017b8aa13045d2 env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
```


`moon ci`

```
[DEBUG 23:45:57] moon_process::command_inspector  Running command git --version  env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:57] moon_process::command_inspector  Running command git branch --show-current  env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[TRACE 23:45:57] log  On default branch master, comparing against previous revision  log.target="moon:query:touched-files" log.module_path="moon_cli::queries::touched_files" log.file="crates/cli/src/queries/touched_files.rs" log.line=52
[DEBUG 23:45:57] moon_process::command_inspector  Running command git merge-base HEAD~1 HEAD  env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:57] moon_process::command_inspector  Running command git merge-base origin/HEAD~1 HEAD  env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:57] moon_process::command_inspector  Running command git merge-base upstream/HEAD~1 HEAD  env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:57] moon_process::command_inspector  Running command git merge-base HEAD HEAD~1  env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
[DEBUG 23:45:57] moon_process::command_inspector  Running command git --no-pager diff --name-status --no-color --relative -z ab89c651dd2dbc57a13e52702de7682eb6f215ff  env_vars={} working_dir="/Users/XXXXX/Documents/XXXXX/XXXXX"
```


It seems that `moon ci` is grabbing the right diff because after all it outputs this

```
[DEBUG 23:45:57] log  Filtering based on touched status "all"  log.target="moon:query:touched-files" log.module_path="moon_cli::queries::touched_files" log.file="crates/cli/src/queries/touched_files.rs" log.line=84
  .drone.yml
  libs/XXXXX/XXXX/main.py
▪▪▪▪ Gathering runnable targets
etc....
````

**Comment by milesj** [comment]
@spljs Just to clarify, the `moon run` is running on a branch correct? So `master` would be the base, and then the branch is trying to merge into master?

We should probably always log the touched files, otherwise its hard to tell what its actually doing 😅

**Comment by spljs** [comment]
Well i'm on master & pushing on master 
But the `moon run --affected` or  `moon run --affected --remote` is not gathering the same files as `moon ci` it seems strange to me as the affected is the same mechanism for `ci` command

**Comment by milesj** [comment]
@spljs Yeah, this won't work right now. `moon ci` is much smarter than `moon run`, as it diffs against the previous commit when on the same branch, `moon run` does not.

I'll look into ways to support this.

**Comment by spljs** [comment]
Alright . Yeah it will be awesome if `moon run` is able to diff against the previous commit . Because with `moon ci` we can not select the tasks to run . It runs every tasks and also don't respect the concurrency . It runs all affected projects in parallel and could cause side effects with some tools

**Comment by spljs** [comment]
Do you have a roadmap on this @milesj  ?

**Comment by milesj** [comment]
@spljs In the next release, so probably next Monday.

**Comment by spljs** [comment]
Awesome thanks !

**Comment by milesj** [comment]
PR: https://github.com/moonrepo/moon/pull/956

**Comment by milesj** [comment]
This has landed in v1.10
